language-plutus-core/src/Language/PlutusCore/Instance/Eq.hs

```diff
@@ -0,0 +1,23 @@
+-- | 'Eq' instances for core data types.
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
```

Comment:
I really don't like orphans. Given it's in the same package, could these instance go with the corresponding types?

Reply:
> I really don't like orphans. 

Yeah, it's a common misconception ([1](https://www.reddit.com/r/haskell/comments/cyfs94/someones_haskell_disappointment_gist_i_came_across/eysjq9t/), [2](https://www.reddit.com/r/haskell/comments/cyfs94/someones_haskell_disappointment_gist_i_came_across/eyrvm94/), [3](https://mail.haskell.org/pipermail/libraries/2008-August/010399.html), [4](https://www.reddit.com/r/haskell/comments/5rcfyd/haskell_maxims_and_arrows/dd6e7sd/), etc). Orphan instances in executables are just fine and there is no good reason to sacrifice modularity for "what if we accidentally forget there already exists an instance and define some another one and use both of those and don't realize that and never import the two files defining the instances in the same file?"

Reply:
> in executables
In `plutus-tx`, ok. This cabal component is a library. I wouldn't mind throwing everything into one bin, makes HIE happy ;-)

> sacrifice modularity
So you want to exchange the `Eq` instances? I've got experience with local coherence in Scala, where you expect it, grudgingly. Would not recommend in Haskell, where it's even lower on your list of things when your code misbehaves.

Reply:
There's no problem with orphans so long as:
- You don't define multiple instances in your package.
- They are always exported along with their datatype.

That's the case here, so I don't see an issue.

The case I think can be risky tends to happen when you define an orphan in separate package from the original type. Then it's plausible that someone else will do the same and import both in different places. That's quite different, though.

Reply:
> This cabal component is a library. 

Err, you're right, I wanted to end this discussion ASAP. Anyway, this still applies:

> there is no good reason to sacrifice modularity for "what if we accidentally forget there already exists an instance and define some another one and use both of those and don't realize that and never import the two files defining the instances in the same file?"

Agree with Michael.

Reply:
> They are always exported along with their datatype.

Ok, sounds good.

Reply:
> They are always exported along with their datatype.

... what does that mean exactly? That the modules defining the datatype should reimport this module?

Reply:
That the modules defining the data type are internal modules not exported outside of the package. The data type gets reexported from a public module together with its instances.

I want to enforce a more strict policy that the orphan instances for a bunch of data types should be in the same folder where the file defining the data type lies. And there should be a file reexporting everything public from that module. See my next PR (some time later).

Reply:
How do orphan instances *actually* go wrong? It happens when you have two modules `A` and `B`, which import a type `T`, and two instances `I1` and `I2` for `T`.

Now, anyone who is going to do this must be able to import `T` without also getting an instance, otherwise they won't be able to define a second one. Orphan instances allow this a priori, because someone could import the module where you've defined the type but not the one where you defined the instance. But this can't happen if you arrange your public package interface so that the instance is always exported when the type is (in the *public* modules).

Of course, this still lets you screw up inside the main package, but it's easy enough to just Not Do That.

Reply:
It will probably make refactoring a bit harder, if you don't know everything that belongs to a datatype - and you might forget a reexport somewhere. I prefer to let the compiler help me anywhere I can.

There are some benefits for [hidden orphans](https://www.parsonsmatt.org/2019/11/27/keeping_compilation_fast.html#hidden-orphans), but I don't think these instances are big enough to profit from the compilation caching.